spectre_analyses/tasks.py

```diff
@@ -57,6 +58,31 @@ def _load_data(dataset_name) -> ty.Dataset:
         return spdata.reader.load_txt(infile)
 
 
+class Cleanup(object):
+    def __init__(self, path: str, old_signal=None):
+        self.path = path
+        self.old = old_signal
+
+    def __call__(self, signal_number, stack_frame):
+        if os.path.exists(self.path):
+            shutil.rmtree(self.path, ignore_errors=True)
+        if self.old is not None:
+            sys.exit(self.old.value)
+        else:
+            sys.exit(0)
+
+    def __repr__(self):
+        return "{0}({1},{2})".format(Cleanup.__name__, self.path, self.old)
+
+
+@contextmanager
+def analysis_cleanup(path: str):
+    cleanup = Cleanup(path)
+    cleanup.old = signal.signal(signal.SIGTERM, cleanup)
+    yield
+    signal.signal(signal.SIGTERM, cleanup.old)
```

Comment:
In fact, what we need here is (1) a signal deferring context manager class, and (2) a descendant cleanup manager class that extends the former with directory removal logic. I’d suggest we have a `defer_signals` class implementing the context manager protocol (`__enter__`, `__exit__`) as written [here](https://gist.github.com/evansd/2375136), and `analysis_cleanup` class that adds cleanup logic with:

```python
class analysis_cleanup(defer_signals):

    def __init__(self, path: str, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.path = path

    def __exit__(self, *args):
        self.cleanup(path)
        super().__exit__(*args)

    def cleanup(self):
        ... rm -r logic here
```

Benefits:
- context manager styles are not mixed (function/class),
- we gain an easily extendable context manager for future use.

Reply:
Additionally, as in [linked implementation](https://gist.github.com/evansd/2375136), signals should be caught and “replayed” to the process after the context manager exits.

Reply:
We do not need signal deferring at all. We just want to execute *single additional action* when a signal comes, and *immediately* go to default handling. Cleanup & signal hijacking is already separated, I see no need for manual implementation of contextmanagers just for sake of contextmanagers.

The only reasonable modification I see here is replaying the signal instead of manual `sys.exit()`.